packages/react/radio-group/src/Radio.tsx

```diff
@@ -39,74 +38,60 @@ const Radio = React.forwardRef((props, forwardedRef) => {
     as = RADIO_DEFAULT_TAG,
     'aria-labelledby': ariaLabelledby,
     name,
-    checked: checkedProp,
-    defaultChecked,
+    checked = false,
     required,
     disabled,
     value = 'on',
-    onCheckedChange,
     ...radioProps
   } = props;
-  const inputRef = React.useRef<HTMLInputElement>(null);
   const [button, setButton] = React.useState<HTMLButtonElement | null>(null);
   const composedRefs = useComposedRefs(forwardedRef, (node) => setButton(node));
-  const buttonSize = useSize(button);
   const labelId = useLabelContext(button);
   const labelledBy = ariaLabelledby || labelId;
-  const [checked = false, setChecked] = useControllableState({
-    prop: checkedProp,
-    defaultProp: defaultChecked,
-  });
+  const hasConsumerStoppedPropagationRef = React.useRef(false);
+  // We set this to true by default so that events bubble to forms without JS (SSR)
+  const isFormControl = button ? Boolean(button.closest('form')) : true;
 
   return (
-    /**
-     * The `input` is hidden from non-SR and SR users as it only exists to
-     * ensure form events fire when the value changes and that the value
-     * updates when clicking an associated label.
-     */
-    <>
-      <input
-        ref={inputRef}
-        type="radio"
-        name={name}
-        checked={checked}
-        required={required}
+    <RadioProvider checked={checked} disabled={disabled}>
+      <Primitive
+        type="button"
+        role="radio"
+        aria-checked={checked}
+        aria-labelledby={labelledBy}
+        data-state={getState(checked)}
+        data-disabled={disabled ? '' : undefined}
         disabled={disabled}
         value={value}
-        style={{
-          position: 'absolute',
-          pointerEvents: 'none',
-          opacity: 0,
-          margin: 0,
-          ...buttonSize,
-        }}
-        onChange={composeEventHandlers(onCheckedChange, (event) => {
-          setChecked(event.target.checked);
+        {...radioProps}
+        as={as}
+        ref={composedRefs}
+        onClick={composeEventHandlers(props.onClick, (event) => {
+          if (isFormControl) {
+            hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
+            // if radio is in a form, stop propagation from the button so that we only propagate
+            // one click event (from the input). We propagate changes from an input so that native
+            // form validation works and form events reflect radio updates.
+            if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
+          }
         })}
```

Comment:
It is a little odd how this is not the source of truth anymore no?

Reply:
you mean because it doesn't call `onCheckedChange` anymore? I thought the same and was waiting to hear your thoughts here because this would literally just do `onCheckedChange(true)`. So, I thought about calling it `onCheck` or something, I dunno... how do you think we should handle this? It's weird.

Reply:
Yeah exactly, I mean when compared with for example `Checkbox` which is driven from the change event on the actual input.